client/app/pages/dashboards/dashboard.js

```diff
@@ -137,8 +137,14 @@ function DashboardCtrl(
     this.extractGlobalParameters();
   });
 
-  const collectFilters = (dashboard, forceRefresh) => {
-    const queryResultPromises = _.compact(this.dashboard.widgets.map((widget) => {
+  const collectFilters = (dashboard, forceRefresh, updatedParameters = []) => {
+    const affectedWidgets = updatedParameters.length > 0 ? this.dashboard.widgets.filter(
+      widget => widget.getQuery() && widget.getQuery().getParametersDefs().some(
+        ({ name }) => updatedParameters.map(p => p.name).includes(name),
+      ),
+    ) : this.dashboard.widgets;
```

Comment:
I'm uncertain `Array.includes` is supported in Edge 11. Better use lodash.

Reply:
Filtering by `name` won't do as they are unique only within a widget scope.
The same widget can show up twice in a dashboard - one with a dashboard-level param, the other with a widget-level param, and with this implementation, both will get refreshed instead of only one. (Same goes for different widgets that have the same param names).

Here's a demo https://deploy-preview-3804--redash-preview.netlify.com/dashboard/param-update-tester. Change the dashboard parameter, and all widgets refresh even though the middle one shouldn't, since it only has a widget-level param.

Reply:
Good catch. I switched it to use the [proper parameter mapping](https://github.com/getredash/redash/pull/3804/commits/283498aebee72397cefefa15ab3c6675b15e2f2d).

Reply:
Works perfectly 👍 

I know you must have worked on making the filter logic readable but I think it's still a bit complex to comprehend. Here's my suggestion:

```js
let affectedWidgets = dashboard.widgets;
if (updatedParameters.length) {
  const updateParamNames = updatedParameters.map(p => p.name);
  affectedWidgets = dashboard.widgets.filter(
    widget => _.chain(widget.getParameterMappings())
      .filter({ type: 'dashboard-level' })
      .some(({ mapTo }) => _.includes(updateParamNames, mapTo))
      .value(),
  );
}
```

Reply:
Chaining prevents webpack from including only the functions we use from lodash :grimacing:

Reply:
@ranbena thanks for introducing me to lodash chain sequences!

I don't think it would be significant to lose tree shaking on lodash, but I'm still not entirely convinced that the change from 
```js
filter({ type }) => type === 'dashboard-level')
```
to
```js
filter({ type: 'dashboard-level' })
```
makes things simpler, especially given that some simpletons such as myself would possibly have to scratch their heads around `_.chain` and `value()`. The main disadvantage I see here is transforming this from an expression (`const affectedWidgets = stuff`) to a declarative block.

I personally find expressions a primary goal for achieving simplicity.

Reply:
> Chaining prevents webpack from including only the functions we use from lodash 

As @rauchy mentioned it does makes things a lot more readable, I wonder if it's really that costly when considering Lodash :grimacing:

Reply:
Forget it! Let's land this already!